### PR TITLE
fix sympy rules, fixes issue #1067

### PR DESCRIPTION
**Pull request by jvanhoefer**
as suggested, goes directly via MathML

### CODE REVIEW

File: python/amici/sbml_import.py [hunk 1]
```diff
@@ -1025,14 +1025,13 @@ def _process_rules(self) -> None:
...
                 for nested_rule in rules:
-                    nested_formula = sp.sympify(
-                        sbml.formulaToL3String(nested_rule.getMath()),
-                        locals=self.local_symbols)
-                    nested_formula = \
-                        str(nested_formula.subs(variable, formula)).replace('**', '^')
-                    if nested_rule.setFormula(nested_formula) != sbml.LIBSBML_OPERATION_SUCCESS:
-                        raise SBMLException(f'Formula {nested_formula} cannot be parsed by libSBML!')
+                    nested_rule_math_ml = sp.printing.print_mathml(
+                        sbml.formulaToL3String(nested_rule.getMath()))
```
Comment (paulstapor) [inline python/amici/sbml_import.py:1029]: Sorry, I'm not sure what `sp.printing.print_mathml` *exactly* does. However, I think the following things are to be considered. If those are fulfilled, things should be fine:
 * replacement of `**` with `^`
 * accounting for local parameters
And I would definitely check the SBML test suite :)
Additionally, if AMICI now can parse additional SBML models, I would account for that in the test-suite, such that we won't support for those any more. :)

Comment (FFroehlich) [reply python/amici/sbml_import.py:1029]: `print_mathml` generates a mathml string that can be correctly parsed by libsbml, s we don't have to perform any acrobatics here to make libsbml understand what we are outputting here.

File: python/amici/sbml_import.py [hunk 2]
```diff
@@ -1025,14 +1025,13 @@ def _process_rules(self) -> None:
...
             if variable in rulevars:
                 for nested_rule in rules:
-                    nested_formula = sp.sympify(
-                        sbml.formulaToL3String(nested_rule.getMath()),
-                        locals=self.local_symbols)
-                    nested_formula = \
-                        str(nested_formula.subs(variable, formula)).replace('**', '^')
-                    if nested_rule.setFormula(nested_formula) != sbml.LIBSBML_OPERATION_SUCCESS:
-                        raise SBMLException(f'Formula {nested_formula} cannot be parsed by libSBML!')
+                    nested_rule_math_ml = sp.printing.print_mathml(
```
Comment (FFroehlich) [inline python/amici/sbml_import.py:1028]: we lost the `nested_formula.subs(variable, formula)` here, which is the key thing this part of the code is supposed to do (flatten out nested rules).

**Comment by FFroehlich** [comment]
There are 3 sbml tests currently failing, please make sure they pass before merging!

**Comment by jvanhoefer** [comment]
I added a test, if `sp.sympify(...).subs(variable, formula)` returns a `sp.Function` object, since this can not be parsed to MathML. This seems to resolve the three SBML tests. @FFroehlich are you fine with this?

**Comment by FFroehlich** [comment]
> I added a test, if `sp.sympify(...).subs(variable, formula)` returns a `sp.Function` object, since this can not be parsed to MathML. This seems to resolve the three SBML tests. @FFroehlich are you fine with this?

Generally okay, as I don't think we want to support every last bit of sbml functionality. However:

- I am not quite sure what this means in terms of #1067. I would assume that this makes it impossible to have piecewise functions in nested rules?
- I think that it's equally problematic if the `sp.Function` is somewhere contained as operator in a `sp.Basic` variable which is currently not covered by the exception. Thus exception handling should probably be implemented more generic.

**Comment by jvanhoefer** [comment]
So, now we substitute stuff with sympy again... Sometimes `sbml.readMathMLFromString(nested_rule_math_ml)` does produce `None` as result, since sympy can not handle the formula of the rule. In this case I throw an `SBMLException`...

If you are fine with this solution now (+SBMLtests pass), I would merge :)

**Comment by FFroehlich** [comment]
Looks good, fine to merge!

**Comment by jvanhoefer** [comment]
We (meaning @LarsFroehling and me ) were actually still looking into this, since it did not work properly on Lars example. It seems, like Lars found the reason for this, but wanted to push the adaptions today/tomorrow

**Comment by jvanhoefer** [comment]
sorry for not making this clear here :/

**Comment by LarsFroehling** [comment]
as @jvanhoefer said I am a bit late unfortunately. 
I think the MathML string needs to be wrapped with a header before it can be read by `readMathMLFromString`

Where would be the best place to commit this?

**Comment by FFroehlich** [comment]
Sorry about that, just open a new PR for this branch.

**Pull request merged**